### ISSUE TITLE
Stable Fetch

### ISSUE THREAD

**Issue by ronag** [nodejs/undici#1737]
What is left for us to consider fetch stable?

**Comment by ronag** [comment]
@KhafraDev @mcollina

**Comment by mcollina** [comment]
I think we would need to make a call on #1203.

**Comment by ronag** [comment]
Performance is not a significant factor, in my opinion, for stability. I think for most use cases, the existing performance is acceptable. If performance is important you shouldn't be using the fetch api to begin with.

**Comment by silverwind** [comment]
> If performance is important you shouldn't be using the fetch api to begin with.

Performance and Convenience don't have to be exclusive. That's like saying if you want performance, you shouldn't use Node.js or JavaScript 😉.

**Comment by ronag** [comment]
> Performance and Convenience don't have to be exclusive.

I'm not saying that.

**Comment by jimmywarting** [comment]
> performance is NOT a MUST for a stable release while correctness IS a MUST.

agree, 

It would be a different if there is a memory leak due to some performence issue or something else then that is a bug and something that needs to be fixed before calling it stable.

**Comment by mcollina** [comment]
I think we should provide a way to solve the cookie issue before calling it stable: https://github.com/nodejs/undici/issues/1437.

**Comment by KhafraDev** [comment]
Until https://github.com/whatwg/fetch/issues/973 is added there's not much we can do about cookie headers, otherwise we'll be stuck with a non-spec compliant method that will eventually be removed in favor of `getSetCookie` anyways. Of course if we implement `getSetCookie` we also have a problem where any changes done to it before it becomes stable will be a breaking change.

>  It's currently stuck on getting implementer interest from two browsers.

This should probably be handled by wintercg

**Comment by mcollina** [comment]
I think fetch without cookies is not really useful.

Apart from that I think we can call in stable in undici and drop the experimental warning in Node.js. I would still leave it experimental there for a bit longer.

**Comment by KhafraDev** [comment]
Once fetch is stable, would changes to the spec that changed behavior be considered breaking? If so, would there be a new major release of undici for otherwise seemingly minor changes? Ref: #1681

**Comment by mcollina** [comment]
We should create a policy for this and possibly share it also with the TSC.

My 2 cents is that approved spec changes _are not breaking_, otherwise Node.js will end up bein significantly behind and we'd end up maintaining too many lines. However we should validate this with the TSC.

**Comment by ronag** [comment]
There is maybe actually on thing that can be blocking here. Blob and the lack of file system backing.

**Comment by KhafraDev** [comment]
undici does support third party Blobs, so you could use fetch-blob for blobs backed up by the file system.

**Comment by ronag** [comment]
But I think it's a problem for FormData? Since then we always use Node blobs... Any reason why we don't just simply use fetch-blob?

**Comment by KhafraDev** [comment]
FormData will use the FileLike class if Blob isn't a node Blob, so both are supported. I think it's the best solution currently - allowing both the global blob and third party ones.

**Comment by ronag** [comment]
Yea, I'm kind of thinking we maybe need https://github.com/nodejs/node/issues/45188 to be done before fetch is "truly stable".

**Comment by mcollina** [comment]
My 2 cents is that this is ready to drop the experimental warning in v19 and v18.

**Comment by mcollina** [comment]
I think we can mark fetch as stable now.

@KhafraDev is there anything missing on your end?

Would you like to get a WPT report published on how spec-complaint this is?

**Comment by KhafraDev** [comment]
WPTs should be ran automatically, the results can be found here: https://wpt.fyi/results/fetch?label=master&label=experimental&product=chrome&product=firefox&product=safari&product=deno&product=node.js&aligned&q=fetch

I don't think there's a way for us to determine how spec compliant we are -
1. we don't run tests in child processes & workers, so at minimum we run half the tests that Deno does (for example). You can see this in the content-encoding tests from the link above, we pass all 8 tests, and Deno passes all 16.
2. we don't run *.window.js tests, even if they might pass, because *.window.js assumes that the environment is a browser (has access to Window, document, etc.)
3. we don't run tests that only work in browsers (CORS, ORB, etc) for obvious reasons.

But if we wanted a "marketing" number (😄) we pass 1282 / 1453 tests we run, or 88.2% pass.

**Comment by Ethan-Arrowood** [comment]
Info like that @KhafraDev is amazing! We can use that in WinterCG in order to adjust the forked fetch spec regarding what is and isn't possible in non-browser environments. Nice!

**Comment by mcollina** [comment]
Any opposed to mark it as stable?

**Comment by ronag** [comment]
SGTM

**Comment by KhafraDev** [comment]
I am, but I haven't had time to write out my concerns. More or less I don't think that node's stability index is a good match for the fetch spec, which changes much more frequently than other specs. I also think it'll make it harder/impossible in some cases to bring bug fixes and new features to older versions of node.

I plan on making a PR to expose undici's WebSocket in node core once undici is updated. I'm assuming there will be a lot of bug reports. If fetch gets a breaking change, we can no longer land bug fixes in node v20 or v18, even if it's unrelated to fetch.

Hypothetically, if fetch was made stable in v18, we wouldn't be able to add WebSockets until v21/22 because undici had a breaking change in one of the v20 releases. Since fetch is "experimental", WebSocket will probably be in one of the v20 releases.

**Comment by Ethan-Arrowood** [comment]
I'd be open to having a discussion about like marking Web APIs as "stable", but they are not enforced by SerVer since they are based on the specs. I.e. Fetch or WebStreams can received breaking changes within a Node.js minor. Just like how a web developer wouldn't necessarily be able to specify exactly which fetch/webstream implementation they want to use. They'd always be subjected to whatever browsers are shipping at that time

All that said, I do believe web spec authors rarely add breaking changes for this exact reason. Can you imagine what would happen if Fetch had a breaking change? The entire web would break overnight.

**Comment by Ethan-Arrowood** [comment]
And like if our implementation is wrong, and we have to fix it in such a way that warrants a "breaking change" I still would classify that as a patch release since its truly a _bug fix_.

**Comment by KhafraDev** [comment]
fetch has actually had at least 2 breaking changes so far:
1. removing `authorization` headers on cross-origin redirects
2. forcing `duplex: 'half'` option in RequestInit when sending a ReadableStream body with fetch

those are the two I remember off the top of my head, but there are probably more I could dig up if I spent an hour looking through the commit history.

**Comment by KhafraDev** [comment]
I wasn't referring to breaking bug fixes, but bug fixes that come after a breaking change. We would have been waiting on v21 for new undici releases if fetch was stable because the 1st breaking change I mentioned above happened in node v20.x (don't remember exactly)

**Comment by mcollina** [comment]
My 2 cents is that we will cut a new major here when fetch() breaks, and then backport bugfixes down to old undici versions on a as-need basis (to being backported in node LTS releases), then go to the trouble of updating undici fetch in those old LTS releases.

This is more semver work for us, but I think it's generically ok.
The alternative is to treat any breaking spec change as a fix (because technically we are not spec compliant anymore) and therefore can land. I think this might also be the sensible thing to do.

What does @nodejs/tsc think?
1. `fetch()` (and all other WHATWG specs)  breaking changes should land only in Node.js and undici semver-majors releases?
2. `fetch()` (and all other WHATWG specs) breaking changes are actually bugfix to rectify a spec compliance, and therefore can go into minor or patch releases?

This has further implications on long term stability of the project.

**Comment by mcollina** [comment]
@KhafraDev Websocket being a new global can only ship in v21, not v20, even if it's included in the bundle.

**Comment by mhdawson** [comment]
In terms of the options, I'd lean towards:

> fetch() (and all other WHATWG specs) breaking changes should land only in Node.js and undici semver-majors releases?

**Comment by aduh95** [comment]
> there's a chance of it landing in v20 since it's under a flag.

`fetch` being stable doesn’t change that AFAIK, new globals can still land on 20.x behind a flag. Or did you mean something else?

**Comment by KhafraDev** [comment]
we couldn't land it if undici had a semver-major change earlier in the v20 cycle, because we wouldn't be able to use the updated version of undici that added it to the fetch bundle

**Comment by mcollina** [comment]
I'm thinking that we would have to maintain quite a few release lines for this to work, and wonder if it would not be more maintainable to move fetch into core itself long term.

**Comment by KhafraDev** [comment]
or re-designate it as experimental 😉

**Comment by mcollina** [comment]
We should signal the ecosystem it is safe to use without breakages :/.

**Comment by Ethan-Arrowood** [comment]
I'm willing to sacrifice our iteration velocity with Fetch in particular for the sake of signaling to the ecosystem that it is safe to use without breakages. 

If a major _issue_ is uncovered, we can fix that within a major no problem (even if its breaking).

If we want to change Fetch in a _major_ way I think it's fine for us to have to wait until the next Node.js major. That said - we could land multiple _major_ Fetch within one upcoming Node.js major right? If let's say we move forward with Undici v5 in Node.js 20 (and subsequently Node.js 21 since that's coming up soon!), then we have two _major_ changes before Node.js 22 comes out... there's nothing wrong with us landing Undici v7 in Node.js 22 right?

**Comment by Ethan-Arrowood** [comment]
This problem also makes me think that Undici has outgrown itself. I've been thinking about migrating this into a monorepo for sometime, and it could really help us out IMO.

We could have an `@undici/core` that just contains the Dispatcher (and other fundamental pieces). Then we have `@undici/fetch` which is dependent on only `@undici/core` (maybe). Now we could go ahead and release `@undici/fetch` independently from `@undici/core`. Land as many breaking change as we want. Fetch users can only rely on that package. Other users could depend on let's say `@undici/request` and wouldn't have to worry as much about the major changes happening to fetch. 

We can automate the whole versioning process with Changesets. Node.js would only need to vendor in the `@undici/fetch` piece. Tests would run faster if we utilized something like Turborepo. Lots of potential pros.

**I don't want to distract from the focus of this thread so I'm going to crosspost this comment to a new issue for additional comments.**

**Comment by KhafraDev** [comment]
> I'm willing to sacrifice our iteration velocity with Fetch in particular for the sake of signaling to the ecosystem that it is safe to use without breakages.

but why...? If we had this philosophy from the start fetch would have been DOA (actually it never would have made it to a point where it could have been integrated into node, unless someone put in 10x the effort I did in undici).

> then we have two major changes before Node.js 22 comes out... there's nothing wrong with us landing Fetch v7 (Undici v7) in Node.js 22 right?

Or we have a breaking change in v23, release undici v6, and node core fetch users are stuck a year waiting for node v24 to become LTS. Repeat that for every time a breaking change occurs. Browsers have a major release every few weeks, maybe up to a month or two. The minor releases fit into this pattern *much* better, where someone will only have to typically wait a few weeks from when there is a bug fix in undici -> being able to use it in node.

> Furthermore, what if we added support to Node.js for super-users where if they bring their own version of Fetch (say a new version of the one from undici), then they can set it to be the global? I mean I think you can already do this like: globalThis.fetch = require('undici').fetch, but we could make it more discrete so its clearer/easier if you want to use the latest version of Fetch from undici -- you can.

Someone else said the same thing, but at this point we might as well just remove fetch from node core and put a note in the docs to use undici. I wouldn't necessarily call fetch a stable feature if there's genuinely a need to monkey patch it.

**Comment by Ethan-Arrowood** [comment]
In your opinion, when could we make Fetch stable then?

**Comment by Ethan-Arrowood** [comment]
We don't have to go full monorepo, but if we separated out the fetch piece from the rest of undici, then we could more easily maintain multiple major version of fetch. Undici can continue to iterate as needed. We can keep multiple major versions of fetch up to date as needed either backporting commits or what not, meanwhile the rest of undici can continue to be stable and iterate. This could be a potential compromise

**Comment by KhafraDev** [comment]
Fetch is stable (in that we very carefully follow the spec and keep it updated), the spec is not (in that they introduce breaking changes far more frequently than other specs). The node stability index does not have enough nuance to differentiate between this. The node release cycle doesn't fit well into how the fetch spec is handled.

Your average spec does fit into this, where breaking changes literally do not happen, ever, out of fear of breaking everyone. Is it heard of for any other WHATWG spec to start throwing errors in once-valid code (talking about `duplex: half` here)? That's why experimental fits into fetch well - new changes can land about as frequently as they do in browsers, and there's no worry about <insert argument from above>.

**Comment by KhafraDev** [comment]
At minimum there needed to be more discussion and I should have blocked the PR until it happened, but I didn't. At least I finally had time to write up most of my thoughts. Even though everyone seems to disagree with me, I think everyone can agree that it being stable causes issues, and keeping it experimental causes an issue.

**Comment by Ethan-Arrowood** [comment]
I'll bring this comment back up: https://github.com/nodejs/undici/issues/1737#issuecomment-1730042327

If Fetch WHATWG Spec is so unlike other WHATWG Specs, that there is enough of a chance its going to introduce _many_ breaking changes over time, then we need to come up with a new classification for Node.js. Something along the lines of "stable, but subjected to non-semver breaking changes due to spec". Maybe `spec-stable`? 

I understand that WHATWG Fetch has introduced breaking changes recently, but is that truly a common occurrence? Are we going to make fetch experimental forever because its spec had a breaking change one time?

**Comment by KhafraDev** [comment]
> I understand that WHATWG Fetch has introduced breaking changes recently, but is that truly a common occurrence? Are we going to make fetch experimental forever because its spec had a breaking change one time?

At least two, since fetch was enabled by default, and a couple PRs open that have breaking changes, and actually another commit that I missed & won't be able to land because fetch is stable now... (https://github.com/whatwg/fetch/commit/5eaaf662dc4fea196c5d32609cc5a13e6ea98bf2, will no longer throw an error on .formData if the content-type header is invalid, afaict).

**Comment by GeoffreyBooth** [comment]
We can be finer-grained than just referring to the entire `fetch` API. Could we declare significant chunks as stable and certain aspects as still experimental? Like if we think the header parsing is still unstable, that could stay experimental while the rest goes stable, something like that.

**Comment by mcollina** [comment]
> We can be finer-grained than just referring to the entire fetch API. Could we declare significant chunks as stable and certain aspects as still experimental? Like if we think the header parsing is still unstable, that could stay experimental while the rest goes stable, something like that.

@GeoffreyBooth what @KhafraDev is proposing is to keep it experimental forever.

**Comment by KhafraDev** [comment]
Documentation only experimental, or add a new stability index like @mcollina or @Ethan-Arrowood suggested.

> fetch() (and all other WHATWG specs) breaking changes are actually bugfix to rectify a spec compliance, and therefore can go into minor or patch releases?

**Comment by GeoffreyBooth** [comment]
> what @KhafraDev is proposing is to keep it experimental forever.

I understand. But do we know where the future changes are likely to occur? As in are there parts of the overall fetch API that we feel confident won't change, that we can call stable while defining other parts as experimental?

**Comment by mcollina** [comment]
I think it's impossible to predict.

**Comment by KhafraDev** [comment]
Formdata (part of the xhr spec and not fetch) and Headers are unlikely to change.

**Comment by mcollina** [comment]
I disagree to keep things experimental forever. This has lead to _massive issues_ in the past, and we should avoid to repeat this mistake.

**Comment by KhafraDev** [comment]
I'd be interested in some backstory about those issues. I don't think fetch should be experimental forever either, I am arguing in favor of a new stability index (currently experimental is better than stable for fetch, so I would prefer it remained experimental since there isn't an alternative).

**Comment by mcollina** [comment]
`async_hooks` being the most prominent, it has been experimental forever, it's relied as production-ready by businesses across the globe.

**Comment by spersico** [comment]
No-one asked me, but I'm putting my 2 cents here:

The options laid out by @mcollina are, IMO, the ones that should be considered:
> 1. `fetch()` (and all other WHATWG specs)  breaking changes should land only in Node.js and undici semver-majors releases?
> 2. `fetch()` (and all other WHATWG specs) breaking changes are actually bugfix to rectify a spec compliance, and therefore can go into minor or patch releases?

An evolution of a spec does not mean that the **current** status of the spec is not respected. If there are no issues with the fetch API as is, and it's compliance with the **current** spec, I don't see why we should consider even inserting a new stability index, when it's the spec, the thing that's actually changing, and not the API.

Keeping this as experimental, is detrimental to Node as a whole because it pushes users to rely on user-space solutions that probably will not be maintained with the same care for stability, performance and security that this is being handled with (_hello, Axios 0.x_).

Lastly, what's stopping adding the stability index after marking this as stable? It's an index that would affect every WHATWG spec, isn't it?

**Comment by KhafraDev** [comment]
the spec changes the api surface.

**Comment by styfle** [comment]
> new changes can land about as frequently as they do in browsers, and there's no worry about <insert argument from above>.

I don’t understand this part. 

Are you saying that browsers regularly ship breaking changes to fetch? 

If so, then it sounds like no one noticed in browsers and therefore wouldn’t notice in Node.js either.

**Comment by spersico** [comment]
> the spec changes the api.

When I say API, I mean the interface of the implementation, not the specification that the implementation should follow (the spec).
An implementation (in this case, `undici`/`fetch`) might choose to follow a specification ([fetch](https://fetch.spec.whatwg.org/)), but that only means choosing to align the implementation's API, with the spec-defined API.

So, in short, you have the spec's defined API and the implementation's API. 

**My point is**: changes in the living spec could hold this forever in the _experimental_ category (because it's a _living_ spec), when the stability of the implementation it's a different and independent thing, that can be handled with SemVer perfectly fine.

**Comment by KhafraDev** [comment]
> Are you saying that browsers regularly ship breaking changes to fetch?

Browsers release new major versions every few weeks/1-2 months. Chrome has 2 releases in October itself, according to [this](https://chromestatus.com/roadmap). I don't know how long it takes to ship in browsers though or if they follow semver, but you could probably expect breaking changes if they implemented the changes to the spec.

> If so, then it sounds like no one noticed in browsers and therefore wouldn’t notice in Node.js either.

People have already noticed breaking changes we made here. If you search for `duplex` or `duplex: 'half'` github's search might be good enough to show you many such issues. I'm assuming it would be harder to find the issue someone reported when we started removing `authorization` headers from cross-origin redirects, but they are noticed.

> that can be handled with SemVer perfectly fine.

The point about a living spec is great and I agree that currently fetch *should* remain experimental. It could follow semver, but it's already causing issues (read further up, there are already changes we cannot merge now that fetch is stable). Is it worth it to have a fetch implementation that isn't compliant with the spec?

> async_hooks being the most prominent

Async_hooks doesn't follow a spec, I don't think it's a fair comparison with fetch.

**Comment by mcollina** [comment]
> Is it worth it to have a fetch implementation that isn't compliant with the spec?

If a user cannot port their code between browsers and servers... fetch is not useful. Ultimately users would need to keep their browsers up-to-date. Therefore we would need to keep fetch in sync with the status quo. 

As a Node.js user, I don't want my application to break because I upgraded Node.js, especially a LTS release. I also think we should strive to have the most compliant implementation possible.

We need to reach a trade-off to the stability required by devs and businesses across the globe.

I'm personally leaning towards considering spec changes minor changes, and ship as usual.
The target use of `fetch()` is isomorphic JS after all, and if the Browsers change, so should we. We want to avoid the case where a code that run on modern browsers cannot run on Node.js LTS releases.

I'm also ok shipping new undici major for each fetch spec breaking change, and then follow the usual process to ask the TSC for approval for landing the breaking change on a case-by-case basis. We know these cases _will_ happen.

**Comment by targos** [comment]
Correct me if I'm wrong, but I think we already shipped "breaking" changes in LTS to the WHATWG URL implementation to follow spec changes.

**Comment by silverwind** [comment]
> If a user cannot port their code between browsers and servers... fetch is not useful

Only trivial cases will be portable anyways. Browsers have CORS and HTTP1 limits which is are not active on server. Server has to worry about proxies (until https://github.com/nodejs/undici/issues/1650), which browsers do not have (because browser proxies are configured transparently from JS viewpoint).

**Comment by KhafraDev** [comment]
If breaking changes can land then ultimately I have no issue, but it still makes little sense to call it stable if breaking changes can land, right? I think a stability index like "living standard" would be much better, stating mostly the same as stable:

> Stability: 2 - Stable. Compatibility with the npm ecosystem is a high priority.
> Stability: ? - Living Standard. Compatibility with the spec is a high priority.

Ultimately I expect that eventually I will be screwed by semver if it's kept stable. Some change can't land because there's some disagreement and then we won't be able to land undici updates for multiple months to a year.

**Comment by jimmywarting** [comment]
One thing i wish NodeJS would do, would be to auto update only the v8 engine in the background. 
Not necessary updating NodeJS core itself and shipping newer features, breaking changes and bug fixes... 
if browser can update their engine without breaking any website, then so should backends?

**Comment by mcollina** [comment]
@jimmywarting I'll hide your comment as off topic for this issue (and my response).
Anyway, the reason why we cannot update V8 version is because Node.js supports binary addons, and we offer ABI guarantees across a major version. You won't be able to experience Node.js as it is right now without it. This fact is mitigated by n-api, which provide an ABI stable layer across Node majors (and multiple runtimes).

**Comment by ronag** [comment]
I see a problem here: people are overusing fetch for non-isomorphic cases instead of using e.g. `undici.request`. In this case the breaking changes can become quite problematic. Making nodejs seem like an "unstable" platform for developing back-end services.

I think it's important that we add more warnings and information to our documentation of the intended use for fetch and what pros/cons it has.

**Comment by Ethan-Arrowood** [comment]
Those same documentation changes need to also live in Node.js' official docs

**Comment by ljharb** [comment]
@ronag this is the exact kind of problem i warned about when node chose to keep the name “fetch”, fwiw. Isomorphic/universal code is not the default, node-specific code is, and authors of that code are the only ones who can safely declare that their usage works universally. At this point, short of renaming fetch in node to something else, documentation updates are the best option remaining afaict.

**Comment by KhafraDev** [comment]
I'm trying to avoid putting more work onto the releasers, and having a new release cycle similar to browsers would not be good.

**Comment by jimmywarting** [comment]
Agree with jaydenseric on https://github.com/nodejs/undici/issues/1737#issuecomment-1767423988.

If Node.js were a browser, people would quickly label it as Safari has become: "Safari is the new IE." Not because of a lack of new web/JavaScript features or coming up with its own ideas, but due to the slow release cycle and the absence of auto-updates. Safari is tied to the macOS releases update, which is rarely updated.

The Node.js release cycle was fine in the beginning, but now that the goal has shifted more towards shipping more and more web standards and following a specification for compliance, the less current release cycles make less sense. Node.js is not automatically updated, so some library authors are frustrated that they still need to support very old V8 versions. It's like still having to support IE10 today.

**Comment by SimenB** [comment]
>some library authors are frustrated that they still need to support very old NodeJS/V8 versions and has to resort to older solution and lots of polyfill & shims for not being able to use the newest and shiniest feature. It's like still having to support IE10 today.

Do you have any examples of projects/authors that are stuck on IE10 level shims/polyfills due to Node's release cycle? `fetch` is already a global in Node, so not sure what's blocking them.

Auto-updating server runtimes seems weird to me - why would you want this?

**Comment by jimmywarting** [comment]
> Do you have any examples of projects/authors that are stuck on IE10 level shims/polyfills due to Node's release cycle? fetch is already a global in Node, so not sure what's blocking them.

I was just talking in general, i did not have any example in mind. and i don't know who is stuck with ie10. it was just a metaphor. i could replace ie10 with phantomjs...

I know some authors who follow/support the LTS release cycles and uses for instance rimraf instead of NodeJS newest rm recursive mode. so they can't remove rimraf b/c they still need to support other runtimes.
gulp / express is very old fashion and never switch to newer stuff - do you know why? cuz developer do not upgrade there NodeJS version. 

if you are building an application then you can pick and choose whatever nodejs version you want, then auto update dose not matter. But if you are building a library and publishing something on npm for others to use. then you can't exactly use new feature that exist in NodeJS v18+

so you can't reliable depend on fetch or fs.openAsBlob cuz you have to support older env. that's what's a #developer-pain

**Comment by SimenB** [comment]
>But if you are building a library and publishing something on npm for others to use. then you can't exactly use new feature that exist in NodeJS v18+

Of course you can, nothing is stopping you from saying what versions of Node your library supports. Node <18 is also EOL, so supporting older versions in new code shouldn't be an automatic choice, unless you think that maintenance burden is fine.

---

This became a bit off topic so I'll stop here, sorry folks 🙏

**Comment by mcollina** [comment]
This is done, closing.
